File: packages/shared/src/types/qrcode.ts
/**
 * 二维码通信协议类型定义
 * 用于冷钱包和热钱包之间的数据交换
 */

/**
 * 二维码数据类型
 */
export enum QRCodeType {
  /** 地址二维码 */
  ADDRESS = 'address',
  /** 签名后的交易 */
  SIGNED_TRANSACTION = 'signed_tx',
  /** 未签名的交易（热钱包 -> 冷钱包） */
  UNSIGNED_TRANSACTION = 'unsigned_tx',
  /** 助记词（仅用于备份恢复） */
  MNEMONIC = 'mnemonic',
}

/**
 * 基础二维码数据结构
 */
export interface QRCodeBase {
  /** 类型 */
  type: QRCodeType;
  /** 版本号 */
  version: string;
  /** 时间戳 */
  timestamp: number;
  /** 链类型 */
  chain: 'eth' | 'btc' | 'sol' | 'bnb' | 'tron';
}

/**
 * 地址二维码数据
 */
export interface AddressQRCode extends QRCodeBase {
  type: QRCodeType.ADDRESS;
  /** 地址 */
  address: string;
  /** 地址标签（可选） */
  label?: string;
}

/**
 * 签名交易二维码数据
 */
export interface SignedTransactionQRCode extends QRCodeBase {
  type: QRCodeType.SIGNED_TRANSACTION;
  /** 签名后的交易数据（JSON字符串） */
  signedTx: string;
  /** 原始交易哈希（用于验证） */
  txHash: string;
}

/**
 * 未签名交易二维码数据（热钱包 -> 冷钱包）
 */
export interface UnsignedTransactionQRCode extends QRCodeBase {
  type: QRCodeType.UNSIGNED_TRANSACTION;
  /** 未签名的交易数据（JSON字符串） */
  unsignedTx: string;
  /** 交易说明 */
  description?: string;
}

/**
 * 助记词二维码数据（仅用于备份恢复，需要加密）
 */
export interface MnemonicQRCode extends QRCodeBase {
  type: QRCodeType.MNEMONIC;
  /** 加密后的助记词 */
  encryptedMnemonic: string;
}

/**
 * 二维码数据联合类型
 */
export type QRCodeData =
  | AddressQRCode
  | SignedTransactionQRCode
  | UnsignedTransactionQRCode
  | MnemonicQRCode;

/**
 * 二维码编码/解码工具
 */
export class QRCodeProtocol {
  private static readonly PROTOCOL_VERSION = '1.0.0';

  /**
   * 编码数据为JSON字符串（用于生成二维码）
   */
  static encode(data: QRCodeData): string {
    const payload = {
      ...data,
      version: this.PROTOCOL_VERSION,
      timestamp: Date.now(),
    };
    return JSON.stringify(payload);
  }

  /**
   * 解码JSON字符串为数据（从二维码扫描）
   */
  static decode(jsonString: string): QRCodeData {
    try {
      const data = JSON.parse(jsonString) as QRCodeData;

      // 验证版本
      if (data.version !== this.PROTOCOL_VERSION) {
        throw new Error(`Unsupported protocol version: ${data.version}`);
      }

      // 验证类型
      if (!Object.values(QRCodeType).includes(data.type)) {
        throw new Error(`Invalid type: ${data.type}`);
      }

      return data;
    } catch (error) {
      throw new Error(`Failed to decode QR code: ${error}`);
    }
  }

  /**
   * 验证二维码数据有效性
   */
  static validate(data: QRCodeData): boolean {
    if (!data.type || !data.chain || !data.version) {
      return false;
    }

    switch (data.type) {
      case QRCodeType.ADDRESS:
        return 'address' in data && typeof data.address === 'string';
      case QRCodeType.SIGNED_TRANSACTION:
        return 'signedTx' in data && 'txHash' in data;
      case QRCodeType.UNSIGNED_TRANSACTION:
        return 'unsignedTx' in data;
      case QRCodeType.MNEMONIC:
        return 'encryptedMnemonic' in data;
      default:
        return false;
    }
  }
}

File: packages/shared/src-tauri/src/qrcode.rs
//! 二维码生成和解析模块（共享）
//! 
//! 用于冷钱包和热钱包之间的数据交换

use serde::{Deserialize, Serialize};
use qrcode::{QrCode, Color};
use image::{RgbImage, Rgb};
use base64::{Engine as _, engine::general_purpose};

/// 二维码数据类型
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum QRCodeType {
    /// 地址二维码
    Address,
    /// 签名后的交易
    SignedTransaction,
    /// 未签名的交易（热钱包 -> 冷钱包）
    UnsignedTransaction,
    /// 助记词（仅用于备份恢复）
    Mnemonic,
}

/// 基础二维码数据结构
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QRCodeBase {
    /// 类型
    #[serde(rename = "type")]
    pub qr_type: QRCodeType,
    /// 版本号
    pub version: String,
    /// 时间戳
    pub timestamp: u64,
    /// 链类型
    pub chain: String,
}

/// 地址二维码数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddressQRCode {
    #[serde(flatten)]
    pub base: QRCodeBase,
    /// 地址
    pub address: String,
    /// 地址标签（可选）
    pub label: Option<String>,
}

/// 签名交易二维码数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SignedTransactionQRCode {
    #[serde(flatten)]
    pub base: QRCodeBase,
    /// 签名后的交易数据（JSON字符串）
    pub signed_tx: String,
    /// 原始交易哈希（用于验证）
    pub tx_hash: String,
}

/// 未签名交易二维码数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnsignedTransactionQRCode {
    #[serde(flatten)]
    pub base: QRCodeBase,
    /// 未签名的交易数据（JSON字符串）
    pub unsigned_tx: String,
    /// 交易说明
    pub description: Option<String>,
}

/// 生成二维码图片（Base64编码）
/// 
/// # Arguments
/// * `data` - 要编码的数据（JSON字符串）
/// * `size` - 二维码尺寸（像素）
/// 
/// # Returns
/// Base64编码的PNG图片数据
pub fn generate_qrcode(data: &str, size: u32) -> Result<String, String> {
    // 生成二维码
    let qr = QrCode::new(data.as_bytes())
        .map_err(|e| format!("Failed to generate QR code: {}", e))?;
    
    // 计算每个模块的像素大小
    let qr_size = qr.width();
    let module_size = size / (qr_size as u32 + 2); // 留出边距
    let image_size = module_size * (qr_size as u32 + 2);
    
    // 创建图像
    let mut img = RgbImage::new(image_size, image_size);
    
    // 填充白色背景
    for pixel in img.pixels_mut() {
        *pixel = Rgb([255, 255, 255]);
    }
    
    // 绘制二维码
    for y in 0..qr_size {
        for x in 0..qr_size {
            let is_dark = matches!(qr[(x, y)], Color::Dark);
            let color = if is_dark {
                Rgb([0, 0, 0]) // 黑色
            } else {
                Rgb([255, 255, 255]) // 白色
            };
            
            // 绘制模块
            let start_x = (x as u32 + 1) * module_size;
            let start_y = (y as u32 + 1) * module_size;
            
            for py in 0..module_size {
                for px in 0..module_size {
                    if start_x + px < image_size && start_y + py < image_size {
                        img.put_pixel(start_x + px, start_y + py, color);
                    }
                }
            }
        }
    }
    
    // 编码为 PNG
    let mut png_bytes = Vec::new();
    {
        let encoder = image::codecs::png::PngEncoder::new(&mut png_bytes);
        #[allow(deprecated)] // encode 方法已废弃，但 write_image 在当前版本不可用
        encoder
            .encode(
                img.as_raw(),
                img.width(),
                img.height(),
                image::ColorType::Rgb8,
            )
            .map_err(|e| format!("Failed to encode PNG: {}", e))?;
    }
    
    Ok(general_purpose::STANDARD.encode(&png_bytes))
}

/// 解析二维码数据（从JSON字符串）
pub fn parse_qrcode_data(json_str: &str) -> Result<serde_json::Value, String> {
    serde_json::from_str(json_str)
        .map_err(|e| format!("Failed to parse QR code data: {}", e))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_qrcode() {
        let data = r#"{"type":"address","version":"1.0.0","timestamp":1234567890,"chain":"eth","address":"0x1234567890123456789012345678901234567890"}"#;
        let result = generate_qrcode(data, 256);
        assert!(result.is_ok());
    }

    #[test]
    fn test_parse_qrcode_data() {
        let data = r#"{"type":"address","version":"1.0.0","timestamp":1234567890,"chain":"eth","address":"0x1234567890123456789012345678901234567890"}"#;
        let result = parse_qrcode_data(data);
        assert!(result.is_ok());
    }
}

